File: js/src/dispatch.ts
/**
 * Tool dispatcher: route a spec tool name + args object to its implementation.
 *
 * The MVP tools route to their handlers; every other tool defined in
 * `spec/tools/` returns `not_implemented` ("<tool> lands in Phase 2").
 * Argument validation is minimal per the spec: required keys must be present
 * (`invalid_args`), and only schema-declared properties are forwarded.
 *
 * Doc state is per process: `call()` uses one module-level Session; the CLI
 * and MCP layers pass their own via `dispatch()`. Errors are thrown as
 * ToolError in-language and serialize as `{"error", "message", "suggestions"}`
 * at the CLI/MCP boundary.
 */
import { ToolError } from "./errors.js";
import { Session } from "./session.js";
import { TOOL_SPECS } from "./specdata/index.js";
import { docxComment } from "./comments.js";
import { docxConvert } from "./convert.js";
import { docxCreate } from "./create.js";
import { docxField } from "./fields.js";
import { docxList } from "./lists.js";
import { docxMedia } from "./media.js";
import { docxRenderPreview } from "./toolsRender.js";
import { docxSection } from "./sections.js";
import { docxStyle, docxFormat } from "./styles.js";
import { docxTable } from "./tables.js";
import { docxTemplateFill } from "./template.js";
import {
  docxDelete,
  docxEditParagraph,
  docxInsert,
  docxReplace,
  docxRevision,
} from "./toolsEdit.js";
import { docxRepair, docxSave, docxValidate } from "./toolsLifecycle.js";
import { docxOpen, docxOutline, docxRead, docxSearch } from "./toolsRead.js";

type ToolFn = (session: Session, args: Record<string, unknown>) => unknown;

const HANDLERS = {
  docx_open: docxOpen,
  docx_outline: docxOutline,
  docx_read: docxRead,
  docx_search: docxSearch,
  docx_replace: docxReplace,
  docx_edit_paragraph: docxEditParagraph,
  docx_insert: docxInsert,
  docx_delete: docxDelete,
  docx_revision: docxRevision,
  docx_validate: docxValidate,
  docx_repair: docxRepair,
  docx_save: docxSave,
  // Phase 2 — stage 1: tables, styles, lists.
  docx_table: docxTable,
  docx_style: docxStyle,
  docx_format: docxFormat,
  docx_list: docxList,
  // Phase 2 — stage 2: comments, sections, media, fields.
  docx_comment: docxComment,
  docx_section: docxSection,
  docx_media: docxMedia,
  docx_field: docxField,
  // Phase 2 — stage 3: templates, create, convert, render adapter.
  docx_template_fill: docxTemplateFill,
  docx_create: docxCreate,
  docx_convert: docxConvert,
  docx_render_preview: docxRenderPreview,
} as unknown as Readonly<Record<string, ToolFn>>;

/** The Phase-1 tool names (everything else in spec/tools/ declines). */
export const MVP_TOOLS: ReadonlySet<string> = new Set(Object.keys(HANDLERS));

const INPUT_SCHEMAS = new Map<string, Record<string, unknown>>(
  TOOL_SPECS.map((t) => [t.name, t.input_schema]),
);

/**
 * Dispatch one tool call against an explicit session (the CLI and MCP layers
 * use this with their process-lifetime session). Throws ToolError.
 */
export function dispatch(session: Session, tool: string, args: unknown = {}): unknown {
  const schema = INPUT_SCHEMAS.get(tool);
  if (schema === undefined) {
    throw new ToolError("not_implemented", `Tool ${tool} is not defined in spec/tools/.`, [
      "See docs/tools/index.md for the tool catalog.",
    ]);
  }
  const fn = HANDLERS[tool];
  if (fn === undefined) {
    throw new ToolError("not_implemented", `${tool} lands in Phase 2 (see ROADMAP.md)`, []);
  }
  const provided = args ?? {};
  if (typeof provided !== "object" || Array.isArray(provided)) {
    throw new ToolError("invalid_args", `${tool}: args must be a JSON object.`, [
      `Check the tool's input_schema in spec/tools/${tool}.json.`,
    ]);
  }
  const argObj = provided as Record<string, unknown>;
  const required = schema["required"];
  const missing = Array.isArray(required)
    ? (required as string[]).filter((key) => !(key in argObj))
    : [];
  if (missing.length > 0) {
    throw new ToolError(
      "invalid_args",
      `${tool}: missing required argument(s): ${missing.join(", ")}.`,
      [`Check the tool's input_schema in spec/tools/${tool}.json.`],
    );
  }
  const properties = schema["properties"];
  const known =
    properties !== null && typeof properties === "object"
      ? new Set(Object.keys(properties))
      : new Set<string>();
  const filtered: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(argObj)) {
    if (known.has(key)) filtered[key] = value;
  }
  return fn(session, filtered);
}

/** The process-wide session backing `call()` and the native Document surface. */
export const defaultSession = new Session();

/**
 * The contract surface (docs/sdks/javascript.md): identical names, JSON
 * shapes, and errors to the Python package and the MCP server.
 */
export async function call(
  tool: string,
  args: Record<string, unknown> = {},
): Promise<Record<string, unknown>> {
  return dispatch(defaultSession, tool, args) as Record<string, unknown>;
}

File: src/docxengine/_dispatch.py
"""Tool dispatcher: route a spec tool name + args object to its implementation.

As of Phase 2 every one of the 24 tools defined in ``spec/tools/`` routes to a
handler — no path returns ``not_implemented`` for a defined tool. The
``not_implemented`` guard is retained for forward-compat: a future spec tool
shipped without a handler still fails cleanly instead of ``KeyError``.
Argument validation is minimal per the spec: required keys must be present
(``invalid_args``), and only schema-declared properties are forwarded.

Doc state is per process: :func:`call` uses one module-level
:class:`~docxengine._session.Session` unless the caller passes its own.
"""

from __future__ import annotations

from collections.abc import Callable, Mapping

from . import (
    _comments,
    _convert,
    _create,
    _fields,
    _lists,
    _media,
    _render,
    _sections,
    _spec,
    _styles,
    _tables,
    _template,
    _tools_edit,
    _tools_lifecycle,
    _tools_read,
)
from ._errors import ToolError
from ._session import Session

_HANDLERS: dict[str, Callable[..., dict[str, object]]] = {
    "docx_open": _tools_read.docx_open,
    "docx_outline": _tools_read.docx_outline,
    "docx_read": _tools_read.docx_read,
    "docx_search": _tools_read.docx_search,
    "docx_replace": _tools_edit.docx_replace,
    "docx_edit_paragraph": _tools_edit.docx_edit_paragraph,
    "docx_insert": _tools_edit.docx_insert,
    "docx_delete": _tools_edit.docx_delete,
    "docx_revision": _tools_edit.docx_revision,
    "docx_validate": _tools_lifecycle.docx_validate,
    "docx_repair": _tools_lifecycle.docx_repair,
    "docx_save": _tools_lifecycle.docx_save,
    "docx_table": _tables.docx_table,
    "docx_style": _styles.docx_style,
    "docx_format": _styles.docx_format,
    "docx_list": _lists.docx_list,
    "docx_comment": _comments.docx_comment,
    "docx_section": _sections.docx_section,
    "docx_media": _media.docx_media,
    "docx_field": _fields.docx_field,
    "docx_template_fill": _template.docx_template_fill,
    "docx_create": _create.docx_create,
    "docx_convert": _convert.docx_convert,
    "docx_render_preview": _render.docx_render_preview,
}

MVP_TOOLS: frozenset[str] = frozenset(_HANDLERS)

_process_session: Session | None = None


def _default_session() -> Session:
    global _process_session
    if _process_session is None:
        _process_session = Session()
    return _process_session


def call(
    tool: str,
    args: Mapping[str, object] | None = None,
    *,
    session: Session | None = None,
) -> dict[str, object]:
    """Invoke ``tool`` with the schema-named ``args``; returns the result object.

    Failures raise :class:`~docxengine._errors.ToolError`; serialize
    ``exc.to_payload()`` at a process boundary.
    """
    if session is None:
        session = _default_session()
    schema = _spec.input_schema(tool)
    if schema is None:
        raise ToolError(
            "not_implemented",
            f"Tool {tool} is not defined in spec/tools/.",
            ["See docs/tools/index.md for the tool catalog."],
        )
    if tool not in _HANDLERS:  # forward-compat guard: unreached for the 24 Phase-2 tools
        raise ToolError("not_implemented", f"{tool} has no registered handler.", [])
    if args is None:
        args = {}
    if not isinstance(args, Mapping):
        raise ToolError(
            "invalid_args",
            f"{tool}: args must be a JSON object.",
            [f"Check the tool's input_schema in spec/tools/{tool}.json."],
        )
    required = schema.get("required", [])
    missing = [key for key in required if key not in args] if isinstance(required, list) else []
    if missing:
        raise ToolError(
            "invalid_args",
            f"{tool}: missing required argument(s): {', '.join(missing)}.",
            [f"Check the tool's input_schema in spec/tools/{tool}.json."],
        )
    properties = schema.get("properties", {})
    known = properties.keys() if isinstance(properties, dict) else ()
    kwargs = {key: value for key, value in args.items() if key in known}
    return _HANDLERS[tool](session, **kwargs)
